src/uct/ib/ud/base/ud_ep.c

```diff
@@ -416,8 +445,263 @@ void uct_ud_ep_process_rx(uct_ud_iface_t *iface, uct_ud_neth_t *neth, unsigned b
 {
-    ucs_debug("UD EP(%p) FLUSH NOT IMPLEMENTED YET", ep);
+    uct_ud_send_skb_t *skb;
+
+    if (uct_ud_ep_ctl_op_check(ep, UCT_UD_EP_OP_CREP)) {
+        skb = uct_ud_ep_prepare_crep(ep);
+    } else if (uct_ud_ep_ctl_op_check(ep, UCT_UD_EP_OP_RESEND)) {
+        skb =  uct_ud_ep_resend(ep);
+    } else if (uct_ud_ep_ctl_op_check(ep, UCT_UD_EP_OP_ACK)) {
+        skb = &iface->tx.skb_inl.super;
+        uct_ud_neth_ctl_ack(ep, skb->neth);
+        uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_ACK);
+    } else if (uct_ud_ep_ctl_op_isnone(ep)) {
+        skb = 0;
+    } else {
+        skb = 0;
+        ucs_fatal("unsupported pending op mask: %x", ep->tx.pending.ops);
+    }
+
+    if (!skb) {
+        /* no pending - nothing to do */
+        return;
+    }
+
+    iface->ops.tx_skb(iface, ep, skb);
+    uct_ud_ep_log_tx(iface, ep, skb);
+    uct_ud_iface_res_skb_put(iface, skb);
+}
+
+/**
+ * pending operations are processed according to priority:
+ * - high prio control:
+ *   - crep reply 
+ *   - resends   
+ * - pending uct requests 
+ * - low prio control: ack reply/ack requests 
+ *
+ * Low priority control can be send along with user data, so
+ * there is a good chance that processing pending uct reqs will 
+ * also deal with the low prio control.
+ * However we can not let pending uct req block control forever.
+ */
+ucs_arbiter_cb_result_t 
+uct_ud_ep_do_pending(ucs_arbiter_t *arbiter, ucs_arbiter_elem_t *elem,
+                     void *arg)
+{
+    uct_ud_ep_t *ep = ucs_container_of(ucs_arbiter_elem_group(elem), 
+                                       uct_ud_ep_t, tx.pending.group);
+    uct_ud_iface_t *iface = ucs_container_of(arbiter, uct_ud_iface_t,
+                                             tx.pending_q);
+    uintptr_t in_async_progress = (uintptr_t)arg;
+    ucs_arbiter_cb_result_t result;
+
+    /* check if we have global resources
+     * - tx_wqe 
+     * - skb 
+     * control messages does not need skb.
+     */
+    if (!uct_ud_iface_can_tx(iface)) {
+        return UCS_ARBITER_CB_RESULT_STOP;
+    }
+
+    /* here we rely on the fact that arbiter
+     * will start next dispatch cycle from the
+     * next group. 
+     * So it is ok to stop if there is no ctl.
+     * However in worst case only one ctl per
+     * dispatch cycle will be send.
+     */
+    if (!uct_ud_iface_has_skbs(iface) && !uct_ud_ep_ctl_op_isany(ep)) {
+        return UCS_ARBITER_CB_RESULT_STOP;
+    }
+
+    /* we can desched group: iff
+     * - no control
+     * - no ep resources (connect or window)
+     **/
+    
+    if (!uct_ud_ep_ctl_op_isany(ep) &&  
+       (!uct_ud_ep_is_connected(ep) ||
+         uct_ud_ep_no_window(ep))) {
+        return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
+    }
+
+    if (&ep->tx.pending.elem == elem) { 
+        uct_ud_ep_do_pending_ctl(ep, iface);
+        if (!uct_ud_ep_ctl_op_isany(ep)) {
+            /* no more ctl - dummy elem can be removed */
+            result = UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
```

Comment:
i think you can avoid the variable "result" and jest return the required status directly.